R/dummy_extract.R

```diff
@@ -0,0 +1,280 @@
+      }
+    }
+
+    add_step(
+      recipe,
+      step_dummy_extract_new(
+        terms = enquos(...),
+        role = role,
+        trained = trained,
+        sep = sep,
+        pattern = pattern,
+        threshold = threshold,
+        other = other,
+        naming = naming,
+        levels = levels,
+        keep_original_cols = keep_original_cols,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_dummy_extract_new <-
+  function(terms, role, trained, sep, pattern, threshold, other, naming, levels,
+           keep_original_cols, skip, id) {
+    step(
+      subclass = "dummy_extract",
+      terms = terms,
+      role = role,
+      trained = trained,
+      sep = sep,
+      pattern = pattern,
+      threshold = threshold,
+      other = other,
+      naming = naming,
+      levels = levels,
+      keep_original_cols = keep_original_cols,
+      skip = skip,
+      id = id
+    )
+  }
+
+#' @export
+prep.step_dummy_extract <- function(x, training, info = NULL, ...) {
+  col_names <- recipes_eval_select(x$terms, training, info)
+
+  if (length(col_names) > 0) {
+    col_names <- check_factor_vars(training, col_names, "step_dummy")
+
+    ## I hate doing this but currently we are going to have
+    ## to save the terms object from the original (= training)
+    ## data
+    levels <- vector(mode = "list", length = length(col_names))
+    names(levels) <- col_names
+    for (col_name in col_names) {
+      elements <- dummy_extract_extract(
+        training[[col_name]], sep = x$sep, pattern = x$pattern
+      )
+
+      lvls <- map(elements, unique)
+      lvls <- unlist(lvls)
+      lvls <- sort(table(lvls), decreasing = TRUE)
+
+      lvls <- lvls[(lvls / length(elements)) > x$threshold]
+      lvls <- names(lvls)
+
+      levels[[col_name]] <- lvls
+    }
+  } else {
+    levels <- NULL
+  }
+
+  step_dummy_extract_new(
+    terms = x$terms,
+    role = x$role,
+    trained = TRUE,
+    sep = x$sep,
+    pattern = x$pattern,
+    threshold = x$threshold,
+    other = x$other,
+    naming = x$naming,
+    levels = levels,
+    keep_original_cols = get_keep_original_cols(x),
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+#' @export
+bake.step_dummy_extract <- function(object, new_data, ...) {
+
+  # If no terms were selected
+  if (length(object$levels) == 0) {
+    return(new_data)
+  }
+
+  col_names <- names(object$levels)
+  keep_original_cols <- get_keep_original_cols(object)
+
+  for (i in seq_along(object$levels)) {
+    orig_var <- names(object$levels)[i]
+
+    elements <- dummy_extract_extract(
+      getElement(new_data, orig_var),
+      sep = object$sep, pattern = object$pattern
+    )
+
+    indicators <- list_to_dummies(elements, sort(object$levels[[i]]), object$other)
+
+    ## use backticks for nonstandard factor levels here
+    used_lvl <- gsub(paste0("^", col_names[i]), "", colnames(indicators))
+    colnames(indicators) <- object$naming(col_names[i], used_lvl)
```

Comment:
See the note about the message in the test cases. 

Should we check for duplicate names before assigning them?

Reply:
This has been fixed with `dummy_extract_names()`